torch/jit/__init__.py

```diff
@@ -724,15 +724,27 @@ def _try_compile_weak_script(fn):
         return entry["compiled_fn"]
 
 
+# ScriptClasses must be new-style classes because we construct them using their
+# __new__ method.
+def _is_new_style_class(cls):
+    if hasattr(cls, '__class__'):
+        return ('__dict__' in dir(cls) or hasattr(cls, '__slots__'))
+
+
 def script(obj, optimize=True, _frames_up=0, _rcb=None):
     if not _enabled:
         return obj
     if _rcb is None:
         _rcb = _jit_internal.createResolutionCallback(_frames_up + 1)
     if inspect.isclass(obj):
-        mod = ScriptClass(obj.__name__)
+        if not _is_new_style_class(obj):
+            raise RuntimeError("TorchScript classes must be new-style classes. Please inherit from 'object'")
+        name = obj.__name__
+        mod = ScriptClass(name)
```

Comment:
Is the ScriptClass class needed? Besides holding the name, it seems like it is just a Module that is passed to _jit_script_class_compile and then discarded. It is confusingly named because it is a Module subtype, not a class.

Reply:
In the future it might be the object that represents the TorchScript class to enable reference semantics, but it doesn't do anything right now. I'll remove